File: kevoree-platform/org.kevoree.platform.osgi.standalone.gui/src/main/scala/org/kevoree/platform/osgi/standalone/gui/KevoreeLeftModel.scala
package org.kevoree.platform.osgi.standalone.gui

import com.explodingpixels.macwidgets._
import javax.swing.{SwingUtilities, JPanel}
import org.kevoree.{ContainerRoot, ContainerNode}
import org.kevoree.framework.aspects.KevoreeAspects._

/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 06/03/12
 * Time: 20:38
 */

class KevoreeLeftModel extends JPanel {

  val model = new SourceListModel()
  val sourceList = new SourceList(model)

  def getSourceList = sourceList


  // sourceList.setColorScheme(new SourceListDarkColorScheme())


  def reload(kmodel: ContainerNode) {
    if (model.getCategories.size() > 0) {
      model.removeCategoryAt(0)
    }

    val category = new SourceListCategory(kmodel.getName)
    val componentItem = new SourceListItem("Components")
    val channelItem = new SourceListItem("Channels")
    val groupItem = new SourceListItem("Groups")
    model.addCategory(category)
    model.addItemToCategory(componentItem, category)
    model.addItemToCategory(channelItem, category)
    model.addItemToCategory(groupItem, category)

    kmodel.getComponents.foreach {
      c =>
        val itc = new SourceListItem(c.getName + ":" + c.getTypeDefinition.getName)
        model.addItemToItem(itc, componentItem)
    }

    kmodel.eContainer.asInstanceOf[ContainerRoot].getHubs.filter(h => h.getRelatedNodes.exists(c => c.getName == kmodel.getName)).foreach {
      c =>
        val itc = new SourceListItem(c.getName + ":" + c.getTypeDefinition.getName)
        model.addItemToItem(itc, channelItem)
    }
    kmodel.eContainer.asInstanceOf[ContainerRoot].getGroups.filter(g => g.getSubNodes.exists(c => c.getName == kmodel.getName)).foreach{g =>
      val itc = new SourceListItem(g.getName + ":" + g.getTypeDefinition.getName)
      model.addItemToItem(itc, groupItem)
    }



    sourceList.useIAppStyleScrollBars()
    //sourceList.setExpanded(category, false)

    SwingUtilities.invokeLater(new Runnable() {
      def run() {
        doLayout()
        repaint()
        revalidate()
      }
    })
  }


}

File: kevoree-platform/org.kevoree.platform.standalone.gui/src/main/java/org/kevoree/platform/standalone/gui/KevoreeLeftModel.java
package org.kevoree.platform.standalone.gui;

import com.explodingpixels.macwidgets.SourceList;
import com.explodingpixels.macwidgets.SourceListCategory;
import com.explodingpixels.macwidgets.SourceListItem;
import com.explodingpixels.macwidgets.SourceListModel;
import org.kevoree.*;
import org.kevoree.framework.kaspects.ChannelAspect;

import javax.swing.*;

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 26/06/13
 * Time: 07:50
 */
public class KevoreeLeftModel extends JPanel {

    private SourceListModel model = new SourceListModel();
    private SourceList _sourceList = new SourceList(model);
    private ChannelAspect channelAspect = new ChannelAspect();

    public SourceList getSourceList() {
        return _sourceList;
    }

    public void reload(ContainerNode kmodel) {
        if (model.getCategories().size() > 0) {
            model.removeCategoryAt(0);
        }
        SourceListCategory category = new SourceListCategory(kmodel.getName());
        SourceListItem componentItem = new SourceListItem("Components");
        SourceListItem channelItem = new SourceListItem("Channels");
        SourceListItem groupItem = new SourceListItem("Groups");
        SourceListItem childItem = new SourceListItem("ChildNodes");
        model.addCategory(category);
        model.addItemToCategory(componentItem, category);
        model.addItemToCategory(channelItem, category);
        model.addItemToCategory(groupItem, category);
        model.addItemToCategory(childItem, category);
        for (ComponentInstance c : kmodel.getComponents()) {
            SourceListItem itc = new SourceListItem(c.getName() + ":" + c.getTypeDefinition().getName());
            model.addItemToItem(itc, componentItem);
        }
        for (Channel c : ((ContainerRoot) kmodel.eContainer()).getHubs()) {
            if (channelAspect.getRelatedNodes(c).contains(kmodel)) {
                SourceListItem itc = new SourceListItem(c.getName() + ":" + c.getTypeDefinition().getName());
                model.addItemToItem(itc, channelItem);
            }
        }
        for (Group g : ((ContainerRoot) kmodel.eContainer()).getGroups()) {
            if (g.getSubNodes().contains(kmodel)) {
                SourceListItem itc = new SourceListItem(g.getName() + ":" + g.getTypeDefinition().getName());
                model.addItemToItem(itc, groupItem);
            }
        }
        for (ContainerNode child : kmodel.getHosts()) {
            SourceListItem itc = new SourceListItem(child.getName() + ":" + child.getTypeDefinition().getName());
            model.addItemToItem(itc, childItem);
        }
        try {
            _sourceList.useIAppStyleScrollBars();

        } catch (Exception e) {

        }
        SwingUtilities.invokeLater(new Runnable() {
            @Override
            public void run() {
                doLayout();
                repaint();
                revalidate();
            }
        });
    }

}
